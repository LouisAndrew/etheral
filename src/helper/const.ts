import { Product } from './schema/product';
import { Order } from './schema/order';

const mockProduct1: Product = {
    pid: '1234',
    name: 'Mock product',
    slug: 'ctg1/product1',
    category: 'ctg1',
    idrPrice: 23000,
    collection: 'fall',
    productImages: [],
    urls: [],
    weight: 20,
    amount: 12,
};

export const mockProducts: Product[] = [mockProduct1];

export const mockOrder: Order = {
    oid: 'asdadwda',
    buyerName: 'Jane Doe',
    buyerEmail: 'jane@doe',
    buyerPhone: 123412412,
    buyerUId: 'admin',
    buyerAddr: 'Storkowerstrasse',
    buyerPostal: 1234143,
    total: 32000,
    currency: 'IDR',
    date: new Date(),
    via: 'va',
    channel: 'bni',
    products: [{ pid: mockProduct1.pid, amount: 12 }],
    paid: false,
    delivered: false,
    transactionData: {
        sessionId: '1234',
        paymentNo: '12',
        paymentName: 'Jane Doe - BNI',
        expired: '1231241412',
        fee: 3000,
    },
    shippingMethod: 'OKE',
};

export const mockOrders: Order[] = [mockOrder];

export const mockOrderShipped: Order = {
    ...mockOrder,
    delivered: true,
    shippingData: {
        shippedDate: new Date(),
        trackingCode: '12342',
        shippedBy: 'admin1',
    },
};
