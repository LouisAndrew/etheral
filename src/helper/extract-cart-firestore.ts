import { intersectionBy } from 'lodash';

import { SetCartArgs } from 'state/actions/cart';
import { InCart } from './schema/firebase-user';
import { Product } from './schema/product';

export default ({
    firestoreCartData,
    allProducts,
}: {
    firestoreCartData: InCart;
    allProducts: Product[];
}): SetCartArgs => {
    const inCartProduct = intersectionBy(allProducts, firestoreCartData, 'pid');
    return inCartProduct.map(product => ({
        product,
        amount: intersectionBy(firestoreCartData, [product], 'pid')[0].amount, // get the intersection between the product -> pid is unique, so the function is guaranteed to return one item.
    }));
};
