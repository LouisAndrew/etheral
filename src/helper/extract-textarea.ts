import { snakeCase } from 'lodash';

/**
 * function to extract options from a textarea
 * @param textarea textarea input (e.g sizes: S, M, L)
 */
export const extractTextArea = (textarea: string) =>
    textarea.split(',').map(text => snakeCase(text).toUpperCase());
