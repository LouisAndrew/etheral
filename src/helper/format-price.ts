export const formatPrice = (price: number | string) => {
    const value =
        typeof price === 'string'
            ? parseFloat(price.replace('IDR ', ''))
            : price;
    return new Intl.NumberFormat('id').format(value);
};
