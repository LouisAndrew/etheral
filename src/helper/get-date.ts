const pad2 = (n: number): string => (n < 10 ? `0${n}` : n.toString());

const getDate = (date: Date) => {
    const year = date.getUTCFullYear();
    const month = pad2(date.getUTCMonth() + 1); // getMonth() is zero-indexed, so we'll increment to get the correct month number
    const day = pad2(date.getUTCDate());
    const hours = pad2(date.getUTCHours() + 7);
    const minutes = pad2(date.getUTCMinutes());

    return year + '-' + month + '-' + day + ' ' + hours + ':' + minutes;
};

const getDateIpaymu = (date: Date) => {
    const year = date.getUTCFullYear();
    const month = pad2(date.getUTCMonth() + 1); // getMonth() is zero-indexed, so we'll increment to get the correct month number
    const day = pad2(date.getUTCDate());
    const hours = pad2(date.getUTCHours() + 7); // UTC offset for WIB => +7
    const minutes = pad2(date.getUTCMinutes());
    const seconds = pad2(date.getUTCSeconds());

    return `${year}${month}${day}${hours}${minutes}${seconds}`;
};

const monthNames = [
    'January',
    'February',
    'March',
    'April',
    'May',
    'June',
    'July',
    'August',
    'September',
    'October',
    'November',
    'December',
];

const getDateReadable = (date: Date) => {
    const year = date.getUTCFullYear();
    const month = date.getUTCMonth();
    const day = date.getUTCDate();

    return `${day} ${monthNames[month]} ${year}`;
};

export { getDate, getDateIpaymu, getDateReadable };
