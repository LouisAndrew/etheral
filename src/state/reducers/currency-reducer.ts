import { CHANGE_CURRENCY } from '../types/currency';

export enum Currencies {
    AUD = 'AUD',
    IDR = 'IDR',
}

export type Action = {
    type: typeof CHANGE_CURRENCY;
    payload: {
        currency: Currencies;
    };
};

export const initialState = {
    currency: Currencies.IDR,
};

export type IState = {
    currency: Currencies;
};

const reducer = (state: IState = initialState, action: Action): IState => {
    const { type } = action;
    if (type === CHANGE_CURRENCY) {
        const {
            payload: { currency },
        } = action;
        return {
            currency,
        };
    } else {
        return state;
    }
};

export default reducer;
