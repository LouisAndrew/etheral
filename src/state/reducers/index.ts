import { combineReducers } from 'redux';
import numberReducer from './number-reducer';
import productReducer from './products-reducer';
import cartReducer from './cart-reducer';
import currencyReducer from './currency-reducer';

export default combineReducers({
    numberReducer,
    productReducer,
    cartReducer,
    currencyReducer,
});
