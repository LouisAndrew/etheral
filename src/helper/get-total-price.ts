import { Product } from 'helper/schema';
import { withDiscount } from './with-discount';
import { Currencies } from 'state/reducers/currency-reducer';

type HelperArgs = {
    [key: string]: any;
    amount: number;
    product: Product;
};

const getTotalPriceIdr = (args: HelperArgs[]) => {
    return args.reduce((acc, curr) => {
        const { amount, product } = curr;

        // destructure prices prop
        const {
            prices: { discountPercentage, idrPrice },
        } = product;

        // get price.
        const price =
            discountPercentage > 0
                ? withDiscount(idrPrice, discountPercentage)
                : idrPrice;

        return acc + amount * price;
    }, 0);
};

const getTotalPriceAud = (args: HelperArgs[]) => {
    return args.reduce((acc, curr) => {
        const { amount, product } = curr;

        // destructure prices prop
        const {
            prices: { discountPercentage, ausPrice },
        } = product;

        // get price.
        const price =
            discountPercentage > 0
                ? withDiscount(ausPrice, discountPercentage)
                : ausPrice;

        return acc + amount * price;
    }, 0);
};

/**
 * function to get price (formatted ?) based on currency
 * @param args product name and prices + its discount percentage.
 * @param currency current currency state
 */
const getPrice = (args: HelperArgs, currency: Currencies) => {
    const {
        product: { prices },
        amount,
    } = args;

    const { discountPercentage, idrPrice, ausPrice } = prices;

    // handle IDR.
    if (currency === Currencies.IDR) {
        const price =
            discountPercentage > 0
                ? withDiscount(idrPrice, discountPercentage)
                : idrPrice;

        return amount * price;
    } else {
        const price =
            discountPercentage > 0
                ? withDiscount(ausPrice, discountPercentage)
                : ausPrice;

        return amount * price;
    }
};

export { getTotalPriceIdr, getTotalPriceAud, getPrice };
