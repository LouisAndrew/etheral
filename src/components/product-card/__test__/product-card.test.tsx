import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

import { mockProduct1 } from 'helper/const';
import ProductCard from '..';
import { Currencies } from 'state/reducers/currency-reducer';

describe('product card component', () => {
    const mockStore = configureStore([]);
    let Element: any;

    beforeEach(() => {
        const store = mockStore({
            currencyReducer: {
                currency: Currencies.IDR,
            },
        });

        store.dispatch = jest.fn();

        Element = (
            <Provider store={store}>
                <ProductCard product={mockProduct1} />;
            </Provider>
        );
    });

    afterEach(cleanup);

    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(Element, div);
    });

    it('it should display an image and display products name and price correctly', () => {
        const { queryByRole, queryByDisplayValue } = render(Element);

        setTimeout(() => {
            const ProductImage = queryByRole('image');
            const ProductName = queryByDisplayValue(mockProduct1.name);
            const ProductPrice = queryByDisplayValue(
                `IDR ${mockProduct1.idrPrice}`
            );

            if (ProductImage) {
                expect(ProductImage).toBeInTheDocument();
            } else {
                fail();
            }

            if (ProductName) {
                expect(ProductName).toBeInTheDocument();
            } else {
                fail();
            }

            if (ProductPrice) {
                expect(ProductPrice).toBeInTheDocument();
            } else {
                fail();
            }
        }, 300);
    });

    it('matches snapshot', () => {
        const tree = renderer.create(Element).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
