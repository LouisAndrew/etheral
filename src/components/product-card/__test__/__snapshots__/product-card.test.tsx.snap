// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`product card component matches snapshot 1`] = `
Array [
  <div
    className="css-vurnku"
  >
    <div
      className="css-12w9nlj"
      data-testid="tile-container"
    >
      <div
        className="css-oryowi"
      >
        <img
          alt="Mock product"
        />
      </div>
    </div>
    <div
      className="css-vurnku"
    >
      Mock product
    </div>
    <div
      className="css-vurnku"
    >
      IDR 
      23000
    </div>
  </div>,
  ";",
]
`;
